service/frontend/workflowHandler.go

```diff
@@ -3098,3 +3124,73 @@ func (wh *WorkflowHandler) bucketName(customBucketName *string) string {
 func (wh *WorkflowHandler) customBucketNameProvided(customBucketName *string) bool {
 	return customBucketName != nil && len(*customBucketName) != 0
 }
+
+func (wh *WorkflowHandler) historyArchived(ctx context.Context, request *gen.GetWorkflowExecutionHistoryRequest, domainID string) bool {
+	if request.GetExecution() == nil || request.GetExecution().GetRunId() == "" {
+		return false
+	}
+	describeMutableStateRequest := &h.DescribeMutableStateRequest{
+		DomainUUID: common.StringPtr(domainID),
+		Execution:  request.Execution,
+	}
+	resp, err := wh.history.DescribeMutableState(ctx, describeMutableStateRequest)
+	return err != nil || resp.GetMutableStateInDatabase() == ""
```

Comment:
Current mental model is DescribeMutableState (or GetMutableState) should fail with EntityNotFoundError.  Current implementation seems to make some weird checks to handle mutableState not deleted from cache scenario.  I would rather fix the actual issue to remove mutable state from cache too during processing of retention timer instead of explicitly checking 'GetMutableStateInDatabase'.

Reply:
Talked in person about this. Got on the same page. I am now deleting the entry from the cache and fixed a bug in Release method of the lru cache. 

The system invariant is whenever we get EntityNotExistsError from GetMutableState we will assume that means history has been archived.